src/cgroups/v2/memory.rs

```diff
@@ -18,7 +24,253 @@ impl Controller for Memory {
 }
 
 impl Memory {
-    fn apply(_: &Path, _: &LinuxMemory) -> Result<()> {
+    fn set<P: AsRef<Path>>(path: P, val: i64) -> Result<()> {
+        if val == 0 {
+            Ok(())
+        } else if val == -1 {
+            common::write_cgroup_file_str(path, "max")
+        } else {
+            common::write_cgroup_file(path, val)
+        }
+    }
+
+    fn apply(path: &Path, memory: &LinuxMemory) -> Result<()> {
+        // if nothing is set just exit right away
+        if memory.reservation.is_none() && memory.limit.is_none() && memory.swap.is_none() {
+            return Ok(());
+        }
+
+        match memory.limit {
+            Some(limit) if limit < -1 => {
+                bail!("invalid memory value: {}", limit);
+            }
+            Some(limit) => match memory.swap {
+                Some(swap) if swap < -1 => {
+                    bail!("invalid swap value: {}", swap);
+                }
+                Some(swap) => {
+                    Memory::set(path.join(CGROUP_MEMORY_SWAP), swap)?;
+                    Memory::set(path.join(CGROUP_MEMORY_MAX), limit)?;
+                }
+                None => {
+                    if limit == -1 {
+                        Memory::set(path.join(CGROUP_MEMORY_SWAP), -1)?;
+                    }
+                    Memory::set(path.join(CGROUP_MEMORY_MAX), limit)?;
+                }
+            },
+            None => {
+                if memory.swap.is_some() {
+                    bail!("unable to set swap limit without memory limit");
+                }
+            }
+        };
+
+        if let Some(reservation) = memory.reservation {
+            if reservation < -1 {
+                bail!("invalid memory reservation value: {}", reservation);
+            }
+            Memory::set(path.join(CGROUP_MEMORY_LOW), reservation)?;
+        }
+
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::cgroups::test::set_fixture;
+    use crate::utils::create_temp_dir;
+    use oci_spec::LinuxMemory;
+    use std::fs::read_to_string;
+
+    #[test]
+    fn test_set_memory_v2() {
```

Comment:
I don't think there is a need to specify v2 in the name. Cargo test shows the complete name of the test e.g. cgroups::v2::memory::test::test_set_memory, so you already know it's v2.

Reply:
makes sense